lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -7111,32 +7111,58 @@ GenericSignature *GenericSignatureBuilder::computeRequirementSignature(
 
 #pragma mark Generic signature verification
 
-void GenericSignatureBuilder::verifyGenericSignature(ASTContext &context,
-                                                     GenericSignature *sig) {
-  llvm::errs() << "Validating generic signature: ";
-  sig->print(llvm::errs());
-  llvm::errs() << "\n";
+void GenericSignatureBuilder::dropAndCompareEachRequirement(
+    ASTContext &context, GenericSignature *baseSig, GenericSignature *sig,
+    bool includeRedundantRequirements, SmallVectorImpl<Requirement> &redundant,
+    SmallVectorImpl<Requirement> &nonRedundant) {
+#ifndef NDEBUG
+  // `baseSig`'s params should be a subset of `sig`, meaning <A where ...>, <A,
+  // B where ...> (respectively) is okay, but the reverse is not.
+  if (baseSig) {
+    auto canSigParams = sig->getCanonicalSignature()->getGenericParams();
+    for (auto baseParam :
+         baseSig->getCanonicalSignature()->getGenericParams()) {
+      assert(llvm::is_contained(canSigParams, baseParam) &&
+             "baseSig params aren't a subset of sig");
+    }
+  }
+#endif
 
-  // Try removing each requirement in turn.
   auto genericParams = sig->getGenericParams();
   auto requirements = sig->getRequirements();
   for (unsigned victimIndex : indices(requirements)) {
-    PrettyStackTraceGenericSignature debugStack("verifying", sig, victimIndex);
+    PrettyStackTraceGenericSignature debugStack("dropping & comparing", sig,
+                                                victimIndex);
 
     // Form a new generic signature builder.
     GenericSignatureBuilder builder(context);
 
-    // Add the generic parameters.
+    // Add the generic parameters, and specifically those from the main
+    // signature, since it may have more
     for (auto gp : genericParams)
       builder.addGenericParameter(gp);
 
-    // Add the requirements *except* the victim.
+    // Add the appropriate requirements
     auto source = FloatingRequirementSource::forAbstract();
-    for (unsigned i : indices(requirements)) {
-      if (i != victimIndex)
-        builder.addRequirement(requirements[i], source, nullptr);
+    // First, the universal requirements from the base
+    if (baseSig) {
+      for (auto req : baseSig->getRequirements())
+        builder.addRequirement(req, source, nullptr);
     }
 
+    // Next, the requirements from the interesting signature.  Sometimes we have
+    // to add all of the leading ones, and others times only those that we've
+    // already decided aren't redundant:
+    auto leadingReqs = includeRedundantRequirements
+                           ? requirements.slice(0, victimIndex)
+                           : nonRedundant;
+    // We definitely have to add all of the remaining trailing ones:
+    auto trailingReqs = requirements.slice(victimIndex + 1);
+
+    for (auto reqs : {leadingReqs, trailingReqs})
```

Comment:
I had no idea you could do this!

Reply:
Neither did I, but it seemed worth trying!